src/components/logout/logout-controller.ts

```diff
@@ -0,0 +1,188 @@
+import { Request, Response } from "express";
+import { decodeJwt, decodeProtectedHeader, jwtVerify } from "jose";
+import { Config } from "../../config";
+import { logger } from "../../logger";
+import { publicJwkWithKidFromPrivateKey } from "../token/helper/key-helpers";
+import {
+  EC_PRIVATE_TOKEN_SIGNING_KEY,
+  EC_PRIVATE_TOKEN_SIGNING_KEY_ID,
+  RSA_PRIVATE_TOKEN_SIGNING_KEY,
+  RSA_PRIVATE_TOKEN_SIGNING_KEY_ID,
+} from "../../constants";
+
+export const logoutController = async (
+  req: Request,
+  res: Response
+): Promise<void> => {
+  const queryParams = req.query;
+  logger.info("Logout request received");
+  const config = Config.getInstance();
+
+  const defaultLogoutUrl = `${config.getSimulatorUrl()}/signed-out`;
+
+  if (Object.keys(queryParams).length === 0) {
+    return res.redirect(defaultLogoutUrl);
+  }
+
+  const idTokenHint = queryParams.id_token_hint;
+  const stateParam = queryParams.state as string | undefined;
+
+  if (!idTokenHint || typeof idTokenHint !== "string") {
+    logger.info("No Id token hint attached, redirecting to default url");
+    return res.redirect(buildRedirectUri(defaultLogoutUrl, stateParam));
+  }
+
+  if (!(await isIdTokenSignatureValid(idTokenHint))) {
+    return res.redirect(
+      buildRedirectUri(defaultLogoutUrl, stateParam, {
+        error_code: "invalid_request",
+        error_description: "unable to validate id_token_hint",
+      })
+    );
+  }
+
+  let parsedIdToken: {
+    clientId: string | undefined;
+    clientSessionId: string | undefined;
+    rpPairwiseId: string | undefined;
+  };
+
+  try {
+    parsedIdToken = parseIdTokenHint(idTokenHint);
+  } catch (error) {
+    // This error is very unlikely to be hit here and is just as unlikley to be hit in
+    // the actual code as we need to first parse the id token hint before
+    // validating it's signature
```

Comment:
Hm - in the actual code, we seem to parse the whole JWT earlier, but here we only seem to parse the header - is that correct?

Reply:
Ah yeah you're right, will try and reflect the actual code a bit more in this bit

Reply:
Ah yeah I have updated the signature verification logic to be a bit more like what the actual code does